File: bindings/csharp/Sop.Examples/VectorSearchAI.cs
using System;
using System.Collections.Generic;
using Sop;

namespace Sop.Examples;

public static class VectorSearchAI
{
    public static void Run()
    {
        Console.WriteLine("\n--- Running Vector Search (AI/RAG Example) ---");
        Console.WriteLine("Scenario: Semantic Product Search");

        using var ctx = new Context();
        var dbOpts = new DatabaseOptions 
        { 
            StoresFolders = new List<string> { "sop_data_vector" },
            Type = (int)DatabaseType.Standalone
        };
        var db = new Database(dbOpts);

        using var trans = db.BeginTransaction(ctx);
        try
        {
            // 1. Open Vector Store
            var vectorStore = db.OpenVectorStore(ctx, "products_vectors", trans);

            // 2. Upsert Embeddings
            // In a real app, these float[] vectors would come from OpenAI/HuggingFace embeddings of the description.
            // Here we simulate:
            // [1, 0, 0] -> Electronics/Computers
            // [0, 1, 0] -> Clothing/Fashion
            // [0, 0, 1] -> Home/Garden

            Console.WriteLine("Indexing products with embeddings...");

            vectorStore.Upsert(new VectorItem 
            { 
                Id = "prod_1", 
                Vector = new float[] { 0.9f, 0.1f, 0.0f }, // Highly "Electronic"
                Payload = new Dictionary<string, object> { { "name", "Gaming Laptop" }, { "price", 1500 } }
            });

            vectorStore.Upsert(new VectorItem 
            { 
                Id = "prod_2", 
                Vector = new float[] { 0.85f, 0.15f, 0.0f }, // Mostly "Electronic"
                Payload = new Dictionary<string, object> { { "name", "Wireless Mouse" }, { "price", 50 } }
            });

            vectorStore.Upsert(new VectorItem 
            { 
                Id = "prod_3", 
                Vector = new float[] { 0.1f, 0.9f, 0.0f }, // Highly "Fashion"
                Payload = new Dictionary<string, object> { { "name", "Leather Jacket" }, { "price", 300 } }
            });

            // 3. Query (Semantic Search)
            // User searches for "computer accessories" -> Embedding might look like [0.88, 0.12, 0.0]
            Console.WriteLine("\nQuerying for 'computer accessories' (simulated vector [0.88, 0.12, 0.0])...");
            
            var queryVector = new float[] { 0.88f, 0.12f, 0.0f };
            var results = vectorStore.Query(queryVector, k: 2);

            foreach (var hit in results)
            {
                // The payload is retrieved automatically!
                var name = hit.Payload["name"];
                var price = hit.Payload["price"];
                Console.WriteLine($"Match: {name} () - Score: {hit.Score:F4}");
            }

            trans.Commit();
        }
        catch
        {
            trans.Rollback();
            throw;
        }
        finally
        {
            // Cleanup
            Console.WriteLine("Cleaning up vector store...");
            try 
            {
                db.RemoveVectorStore(ctx, "products_vectors");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Cleanup failed: {ex.Message}");
            }
        }
    }
}

File: bindings/rust/examples/vector_search_ai.rs
use sop::{Context, Database, DatabaseOptions, VectorItem, VectorQueryOptions, DatabaseType, L2CacheType};
use std::collections::HashMap;
use serde_json::json;

fn main() {
    println!("\n--- Running Vector Search (AI/RAG Example) ---");
    println!("Scenario: Semantic Product Search");

    let ctx = Context::new();
    if let Some(err) = ctx.error() {
        eprintln!("Error creating context: {}", err);
        return;
    }

    let mut options = DatabaseOptions::default();
    options.stores_folders = Some(vec!["sop_data_vector".to_string()]);
    options.db_type = DatabaseType::Standalone;
    options.cache_type = L2CacheType::InMemory;

    let db = match Database::new(&ctx, options) {
        Ok(db) => db,
        Err(e) => {
            eprintln!("Error creating database: {}", e);
            return;
        }
    };

    let trans = db.begin_transaction(&ctx).unwrap();

    // 1. Open Vector Store
    let vector_store = match db.open_vector_store(&ctx, "products_vectors", &trans) {
        Ok(vs) => vs,
        Err(e) => {
            eprintln!("Error opening vector store: {}", e);
            return;
        }
    };

    // 2. Upsert Embeddings
    println!("Indexing products with embeddings...");

    let mut payload1 = HashMap::new();
    payload1.insert("name".to_string(), json!("Gaming Laptop"));
    payload1.insert("price".to_string(), json!(1500));

    let item1 = VectorItem {
        id: "prod_1".to_string(),
        vector: vec![0.9, 0.1, 0.0],
        payload: payload1,
    };

    if let Err(e) = vector_store.upsert(&ctx, item1) {
        eprintln!("Error upserting item 1: {}", e);
    }

    let mut payload2 = HashMap::new();
    payload2.insert("name".to_string(), json!("Wireless Mouse"));
    payload2.insert("price".to_string(), json!(50));

    let item2 = VectorItem {
        id: "prod_2".to_string(),
        vector: vec![0.85, 0.15, 0.0],
        payload: payload2,
    };

    if let Err(e) = vector_store.upsert(&ctx, item2) {
        eprintln!("Error upserting item 2: {}", e);
    }

    trans.commit(&ctx).unwrap();
    println!("Indexing complete.");

    // 3. Search
    let trans2 = db.begin_transaction(&ctx).unwrap();
    
    println!("Searching for 'electronic devices' (simulated vector [0.9, 0.0, 0.0])...");
    
    let query = VectorQueryOptions {
        vector: vec![0.9, 0.0, 0.0],
        k: 2,
        filter: None,
    };

    match vector_store.search(&ctx, query) {
        Ok(results) => {
            for res in results {
                println!("Found: {} (Score: {:.4}) - Payload: {:?}", res.id, res.score, res.payload);
            }
        },
        Err(e) => eprintln!("Error searching: {}", e),
    }

    trans2.commit(&ctx).unwrap();
}
